fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/ReindexWorker.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed to DuraSpace under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional information
+ * regarding copyright ownership.
+ *
+ * DuraSpace licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except in
+ * compliance with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.fcrepo.persistence.ocfl.impl;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A reindexing worker thread.
+ * @author whikloj
+ */
+public class ReindexWorker implements Runnable {
+
+    private Thread t;
+    private ReindexManager manager;
+    private ReindexService service;
+    private boolean running = true;
+    private boolean failOnError;
+
+    /**
+     * Basic Constuctor
+     * @param reindexManager the manager service.
+     * @param reindexService the reindexing service.
+     * @param failOnError whether the thread should fail on an error or log and continue.
+     */
+    public ReindexWorker(final ReindexManager reindexManager, final ReindexService reindexService,
+                         final boolean failOnError) {
+        manager = reindexManager;
+        service = reindexService;
+        this.failOnError = failOnError;
+        t = new Thread(this, "ReindexWorker");
+    }
+
+    /**
+     * Join the thread.
+     * @throws InterruptedException if the current thread is interrupted.
+     */
+    public void join() throws InterruptedException {
+        t.join();
+    }
+
+    /**
+     * Start the thread with this Runnable
+     */
+    public void start() {
+        t.start();
+    }
+
+    @Override
+    public void run() {
+        while (running) {
+            final List<String> ids = manager.getIds();
+            if (ids.isEmpty()) {
+                stopThread();
+                break;
+            }
+            final Map<String, String> states = new HashMap<>();
+            for (final var id : ids) {
+                try {
+                    service.indexOcflObject(manager.getTransactionId(), id);
+                    states.put(id, "");
+                } catch (final Exception e) {
+                    states.put(id, e.getMessage());
+                    if (failOnError) {
```

Comment:
The problem with this behavior is that it will only ever stop _this_ thread on error. The rest of the threads will proceed. One solution to this problem is to change the status reporting, as previously discussed, and have the `updateComplete` method trigger stopping all of the threads when an error is reported and `failOnError` is true.

Reply:
Because each worker has a reference to the ReindexManager, I was just going to add a call to `ReindexManager.stop()` which stops all the threads. Do you see any problems there?

Reply:
Yes, that should be fine